### PR TITLE
Sibling Doublets for the HGCal Pattern Recoginition by CA

### PR DESCRIPTION
**Pull request by waredjeb**
#### PR description:

This pull request introduces the Sibling Doublets for the HGCal Pattern Recognition. 
This feature has been already discussed in the last Reconstruction and Analysis Tools meeting ([link to presentation](https://indico.cern.ch/event/1095576/contributions/4608272/attachments/2345071/3998685/HGCAL%20Reco%20meeting%20November%202021.pdf), slide 14)
- Some validation plot [here](https://wredjeb.web.cern.ch/HGCAL/SiblingsDoublets/New/D76/), tiny and understood changes in physics
- Timing and memory consumption evaluated on a pileup 200 workflow (34902.999)

### This PR
```
FastReport --------------------------- Run 1 Summary ---------------------------
FastReport   CPU time avg.      when run  Real time avg.      when run     Alloc. avg.      when run   Dealloc. avg.      when run  Modules
FastReport        47.7 ms        47.7 ms        47.8 ms        47.8 ms      +26779 kB      +26779 kB      -13708 kB      -13708 kB    ticlLayerTileProducer
FastReport         0.0 ms         0.0 ms         0.0 ms         0.0 ms          +0 kB          +0 kB          +0 kB          +0 kB    ticlSeedingGlobal
FastReport         4.5 ms         4.5 ms         4.5 ms         4.5 ms         +53 kB         +53 kB         -26 kB         -26 kB    ticlSeedingTrk
FastReport         1.1 ms         1.1 ms         1.2 ms         1.2 ms      +10846 kB      +10846 kB       -9821 kB       -9821 kB    ticlSimTracksters
FastReport       822.9 ms       822.9 ms       822.9 ms       822.9 ms     +239437 kB     +239437 kB     -237668 kB     -237668 kB    ticlTrackstersEM
FastReport        97.0 ms        97.0 ms        97.1 ms        97.1 ms      +49810 kB      +49810 kB      -48358 kB      -48358 kB    ticlTrackstersHAD
FastReport       435.9 ms       435.9 ms       435.9 ms       435.9 ms      +72455 kB      +72455 kB      -69580 kB      -69580 kB    ticlTrackstersMerge
FastReport       678.0 ms       678.0 ms       678.0 ms       678.0 ms     +163675 kB     +163675 kB     -161092 kB     -161092 kB    ticlTrackstersTrk
FastReport       520.1 ms       520.1 ms       523.4 ms       523.4 ms     +188421 kB     +188421 kB     -186847 kB     -186847 kB    ticlTrackstersTrkEM

Total Time = 2610.3 ms
```

```
### Release
FastReport --------------------------- Run 1 Summary ---------------------------
FastReport   CPU time avg.      when run  Real time avg.      when run     Alloc. avg.      when run   Dealloc. avg.      when run  Modules
FastReport        46.5 ms        46.5 ms        46.5 ms        46.5 ms      +26779 kB      +26779 kB      -13708 kB      -13708 kB    ticlLayerTileProducer
FastReport         0.0 ms         0.0 ms         0.0 ms         0.0 ms          +0 kB          +0 kB          +0 kB          +0 kB    ticlSeedingGlobal
FastReport         4.5 ms         4.5 ms         4.5 ms         4.5 ms         +53 kB         +53 kB         -26 kB         -26 kB    ticlSeedingTrk
FastReport         1.1 ms         1.1 ms         1.1 ms         1.1 ms      +10846 kB      +10846 kB       -9821 kB       -9821 kB    ticlSimTracksters
FastReport       944.4 ms       944.4 ms       944.5 ms       944.5 ms     +349877 kB     +349877 kB     -348249 kB     -348249 kB    ticlTrackstersEM
FastReport       100.5 ms       100.5 ms       100.6 ms       100.6 ms      +60440 kB      +60440 kB      -59009 kB      -59009 kB    ticlTrackstersHAD
FastReport       446.8 ms       446.8 ms       446.9 ms       446.9 ms      +72579 kB      +72579 kB      -70559 kB      -70559 kB    ticlTrackstersMerge
FastReport       719.3 ms       719.3 ms       719.3 ms       719.3 ms     +176474 kB     +176474 kB     -174532 kB     -174532 kB    ticlTrackstersTrk
FastReport       534.6 ms       534.6 ms       535.7 ms       535.7 ms     +205419 kB     +205419 kB     -203905 kB     -203905 kB    ticlTrackstersTrkEM

Total Time =  2799.3ms
```
@rovere  @felicepantaleo @cseez @pfs

### CODE REVIEW

File: RecoHGCal/TICL/plugins/HGCGraph.cc [hunk 1]
```diff
@@ -221,6 +238,16 @@ bool HGCGraphT<TILES>::areTimeCompatible(int innerIdx,
           std::abs(timeIn - timeOut) < maxDeltaTime * sqrt(timeInE * timeInE + timeOutE * timeOutE));
 }
 
+template <typename TILES>
+bool HGCGraphT<TILES>::areOverlappingOnSiblingLayers(int innerIdx,
+                                                     int outerIdx,
+                                                     const std::vector<reco::CaloCluster> &layerClusters,
+                                                     float maxRSquared) {
+  float etaDiff = layerClusters[outerIdx].eta() - layerClusters[innerIdx].eta();
+  float phiDiff = layerClusters[outerIdx].phi() - layerClusters[innerIdx].phi();
```
Comment (jpata) [inline RecoHGCal/TICL/plugins/HGCGraph.cc:247]: will this bring wraparound issues? maybe better to use a deltaR function?

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoHGCal/TICL/plugins/HGCGraph.cc (hunk 1 above)
Comment (waredjeb) [reply RecoHGCal/TICL/plugins/HGCGraph.cc]: Right! replaced with deltaR2 method

File: RecoHGCal/TICL/plugins/PatternRecognitionbyCA.cc [hunk 2]
```diff
@@ -184,6 +187,13 @@ void PatternRecognitionbyCA<TILES>::makeTracksters(
       tmp.setSeed(input.regions[0].collectionID, seedIndices[tracksterId]);
 
       std::copy(std::begin(effective_cluster_idx), std::end(effective_cluster_idx), std::back_inserter(tmp.vertices()));
+      // Propagate the correct graph connections
+      tmp.edges().reserve(ntuplet.size());
+      for (auto const &t : ntuplet) {
+        std::array<unsigned int, 2> edge = {
+            {(unsigned int)doublets[t].innerClusterId(), (unsigned int)doublets[t].outerClusterId()}};
```
Comment (slava77) [inline RecoHGCal/TICL/plugins/PatternRecognitionbyCA.cc:194]: from 4.15 http://cms-sw.github.io/cms_coding_rules.html#4--technical-coding-rules-1
```suggestion
            {static_cast<unsigned int>(doublets[t].innerClusterId()), static_cast<unsigned int>(doublets[t].outerClusterId())}};
```

File: RecoHGCal/TICL/python/EMStep_cff.py [hunk 3]
```diff
@@ -29,6 +29,7 @@
         shower_start_max_layer = 5, #inclusive
         max_out_in_hops = 1,
         skip_layers = 2,
+        siblings_maxRSquared = [6e-4, 6e-4, 6e-4],
```
Comment (slava77) [inline RecoHGCal/TICL/python/EMStep_cff.py:32]:
```suggestion
```
not needed here (the same as the default already).

Same comment for the other files below

**Comment by apsallid** [comment]
Hi @waredjeb , 

If @rovere and @felicepantaleo agree and since you are in there please correct `outerR()` [here](https://github.com/cms-sw/cmssw/blob/master/RecoHGCal/TICL/plugins/HGCDoublet.h#L52). I understand that it isn't used anywhere but it should return `outerR_` not `outerZ_`. Sorry for the out of topic!

File: RecoHGCal/TICL/plugins/PatternRecognitionbyCA.cc [hunk 4]
```diff
@@ -184,6 +187,13 @@ void PatternRecognitionbyCA<TILES>::makeTracksters(
       tmp.setSeed(input.regions[0].collectionID, seedIndices[tracksterId]);
 
       std::copy(std::begin(effective_cluster_idx), std::end(effective_cluster_idx), std::back_inserter(tmp.vertices()));
+      // Propagate the correct graph connections
+      tmp.edges().reserve(ntuplet.size());
+      for (auto const &t : ntuplet) {
+        std::array<unsigned int, 2> edge = {{static_cast<unsigned int>(doublets[t].innerClusterId()),
+                                             static_cast<unsigned int>(doublets[t].outerClusterId())}};
+        tmp.edges().push_back(edge);
```
Comment (rovere) [inline RecoHGCal/TICL/plugins/PatternRecognitionbyCA.cc:195]: Did you check the increase in the size of the `Trackster` collection?

File: RecoHGCal/TICL/plugins/HGCDoublet.cc [hunk 5]
```diff
@@ -35,14 +37,18 @@ bool HGCDoublet::checkCompatibilityAndTag(std::vector<HGCDoublet> &allDoublets,
...
+      }
       ok[j] = areAligned(xi[j], yi[j], zi[j], xo, yo, zo, minCosTheta, minCosPointing, refDir, debug);
       if (debug) {
         LogDebug("HGCDoublet") << "Are aligned for InnerDoubletId: " << i + j << " is " << ok[j] << std::endl;
       }
     }
     for (int j = 0; j < vs; ++j) {
       auto otherDoubletId = innerDoublets[i + j];
-      auto &otherDoublet = allDoublets[otherDoubletId];
+      auto &otherDoublet = allDoublets.at(otherDoubletId);
```
Comment (rovere) [inline RecoHGCal/TICL/plugins/HGCDoublet.cc:51]: Why did we move to use `at` with its underlying boundary check?

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
There is a fairly significant increase in the number of PF electrons
mostly with `|eta|>2.2` in D77 workflow with PU (35234.999)
all_OldVSNew_TTbar14TeV2026D77PUwPRMXwf35234p999c_recoPFCandidates_particleFlow__RECO_obj_eta61
while the number of photons goes down
all_OldVSNew_TTbar14TeV2026D77PUwPRMXwf35234p999c_recoPFCandidates_particleFlow__RECO_obj_eta75

is this expected?

The validation plots linked in the PR description do not show performance in the higher level quantities.

@cms-sw/egamma-pog-l2 @cms-sw/pf-l2 
please take a note and clarify if there are some concerns to merge this PR or if it should be OK to merge and deal with any updates downstream (if needed) in follow up PRs.

**Comment by swagata87** [comment]
Hi,
- Do I understand correctly that its only the PF electrons / PF photons where we see significant changes, but the GED electrons / GED photons do not show any significant change? 
- Was there any study made on PU=0 sample? If yes, do we see differences for PF ele/pho in PU=0 sample as well? 
- Number of PF photon went down, in ttbar sample. As ttbar sample does not have real photons, this could mean that jet-faking-photon rate has decreased. If that's the case, then it's fine. But is there any study supporting it?
- Apart from that one slide in the RECO meeting talk, is there any other presentation on this topic that I can look at, to know more about the motivation of this PR and the changes expected for superclusters, electrons and photons?

**Comment by slava77** [comment]
@waredjeb @rovere 
please check https://github.com/cms-sw/cmssw/pull/36171#issuecomment-977204572 and clarify on the physics expectations in this PR.
Thank you.

**Comment by rovere** [comment]
@slava77 thanks for the reminder!
@waredjeb is working on this.

**Comment by felicepantaleo** [comment]
enable profiling

**Comment by felicepantaleo** [comment]
@cmsbuild please test

**Comment by swagata87** [comment]
@waredjeb has made some further studies on photons (he can link his plots here, for record). 
Fakerate of GED photons got significantly reduced after this PR. However, there is a small cost of efficiency. The observation of less number of reconstructed photons after this PR is mostly driven by reduced fakerate. 

Given the improved fakerate of photons, and also gain in timing[1], the small efficiency loss seems tolerable to me.  

[1]
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-af02b3/20966/profiling/23434.21/RES_CPU_compare_23434.21.txt
```
[3    -> 4   ] [4.2    -> 3.6   ]  [58.32     -> 49.54    ]         [ 49.54    /1372.46   - 58.32    /1374.99   *100% ] = [-0.63 % ]         TrackstersProducer::produce  
```

**Comment by slava77** [comment]
+reconstruction

for https://github.com/cms-sw/cmssw/pull/36171	0e21ba7
- code changes are in line with the PR description and the follow up review; including the more detailed physics check as summarised in https://github.com/cms-sw/cmssw/pull/36171#issuecomment-986299176
- jenkins tests pass and comparisons with the baseline show differences in phase-2 workflows starting in HGCAL/trackster objects and propagating downstream
- the profiling plots show around 0.8% reduction in CPU time, driven mainly by the speedup in TrackstersProducer::produce

**Comment by srimanob** [comment]
@waredjeb @swagata87 
Would you mind putting your study plot here for reference? Thanks.

**Comment by waredjeb** [comment]
Hi @srimanob, sure. Here you can find the new plots https://wredjeb.web.cern.ch/HGCAL/PRTest/plots/, workflow 35252.0 PU200. 
Some additional information for these plots:
- I rebased my branch to include #36204 
- Selection on Particle Flow photons: pT > 10GeV (same applied on GenParticles)

The photons fake rate increases a bit with this PR. We had an offline discussion with Soham, who said that given the CPU time reduction, the small changes are fine.

**Comment by slava77** [comment]
@cms-sw/upgrade-l2 
please check if this PR needs further info or if it can be signed soon.

**Comment by srimanob** [comment]
+Upgrade

Slightly change of photon fake rate is expected from this PR.

**Comment by qliphy** [comment]
+1

**Pull request merged**